### PR TITLE
native: Resurect native/perif/timer and handle #715 (WIP)

### PR DESCRIPTION
**Pull request by benoit-canet**
-An eINT() was missing before calling the trampoline.
-A potential dINT()/eINT() recursive call has been eliminated.
-schedule_timer() was setting the native timer while signals where disabled this could leads to SIGALARM being blocked.
-Add a define for the native platform timer resolution
-A bunch of case in the signal handler where mishandled leading to usefull signal being not processed.So a mechanism to postpone the processing of these usefull signals was added.

Pass the hwtimer* and vtimer* tests.

### CODE REVIEW

**Comment by benoit-canet** [comment]
There is still a very rare early random crash of hwtimer_wait() I will look at it.

File: cpu/native/irq_cpu.c [hunk 1]
```diff
@@ -325,6 +385,7 @@ void native_isr_entry(int sig, siginfo_t *info, void *context)
     _native_cur_ctx = (ucontext_t *)sched_active_thread->sp;
 
     DEBUG("\n\n\t\tnative_isr_entry: return to _native_sig_leave_tramp\n\n");
+    dINT();
```
Comment (haukepetersen) [inline cpu/native/irq_cpu.c]: please use `enableIRQ()` instead of the deprecated `dINT`

**Comment by haukepetersen** [comment]
Nice one, pulling this into my network stack test branch prevents some ISR error messages at the least... Let us know once we can remove the WIP and test this in more depth!

**Comment by kaspar030** [comment]
@benoit-canet Do you mind also fixing #2870? That periph timer implementation is basically using the same code, and I think you could adapt the fixes a lot quicker than me.

**Comment by benoit-canet** [comment]
@kaspar030 could #2870 be implemented on top of native hwtimers ?
I am putting some effort resurecting it and would like the code be reused.

**Comment by benoit-canet** [comment]
Ah !

I just saw you created RIOT.

I think I will finish to polish hwtimer first by fixing corner case bugs in native.
Then I'll probably move to #2870.

**Comment by haukepetersen** [comment]
Just one idea: does it make sense to switch over to the `periph_timer` entirely on native and use the `hwtimer_compat` module on top of that for now?

**Comment by kaspar030** [comment]
> does it make sense to switch over to the periph_timer entirely on native and use the hwtimer_compat module on top of that for now?

Totally, as it would bring native closer to many other platforms, and get the periph/timer implementation some testing until we switch away from hwtimer/vtimer completely.

**Comment by haukepetersen** [comment]
That's what I wanted to hear :-)

**Comment by kaspar030** [comment]
@benoit-canet hwtimer will be phased out soon, as we think the slimmer periph/timer interface coupled with a redesigned higher level interface is the way to go.

#2870 is basically copy&paste from native hwtimer code to implement the periph/timer interface. It actually makes more sense to concentrate on stabilizing that.

Compare #2926 to see where we're headed.

**Comment by benoit-canet** [comment]
Covered a few more corner cases that could cause hwtimer_wait stall.

There is still a rare intermitent crash but I think it related on how context switchs are done
on the native platform: the cpu try to jump to a stack start...

I test with:

--#!/bin/bash

make clean all-debug

while :
do

screen -d -m -L gdb -ex 'run > log' ./bin/native/hwtimer_wait.elf
pid=$!
sleep 9

success=$(cat log|grep success)

if [ "$success" != "success"  ]
then
echo "\nBUG\n"
exit
fi

pkill screen
pkill hwtimer

done
--

**Comment by benoit-canet** [comment]
@kaspar030 I'll look a these issues/PR.

**Comment by benoit-canet** [comment]
@kaspar030 I resurected your perif/timer series an rebased it in top of my patches.

**Comment by benoit-canet** [comment]
@kaspar030 

Why the idle_thread never try to thread_yield_higher ?

What if it's stuck in lpm's pause and the signal that wake it up does not reschedule something else ?

**Comment by kaspar030** [comment]
> Why the idle_thread never try to thread_yield_higher ?

In RIOT, a thread is always run if there's no higher thread waiting to be run. 

CPU time is never time-sliced between multiple running threads, the scheduler is completely tickless.
This implies that if the idle_thread gets scheduled, all other threads are in some sleep mode.

> What if it's stuck in lpm's pause and the signal that wake it up does not reschedule something else ?

Interrupts are supposed to trigger the scheduler, i.e., by setting "sched_context_switch_request = 1", if another thread might need scheduling. So if idle goes to LPM, then an interrupt occurs, and another thread needs to be run, the scheduler would switch to this other thread, and only if this other thread (and all others) are finished, switch back to the idle thread.

**Comment by benoit-canet** [comment]
@kaspar030 thanks for the clear explanation.

**Comment by benoit-canet** [comment]
@kaspar030 work fine on two machines now up for review

**Comment by benoit-canet** [comment]
I don't have access to an osX machine for testing though.

**Comment by kaspar030** [comment]
@benoit-canet I've tried the ng_networking example. It loses some packets (compared to master, which loses none). Also, when I hammer it with pings from the linux box, it crashes at some point:

```
[kaspar@booze ng_networking (benoit-timer)]$ /home/kaspar/src/riot.8/examples/ng_networking/bin/native/ng_networking.elf tap0
RIOT native uart0 initialized.
RIOT native interrupts/signals initialized.
LED_GREEN_OFF
LED_RED_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

kernel_init(): This is RIOT! (Version: 2014.12-2057-g2f60e5-booze-benoit-timer)
kernel_init(): jumping into first task...
UART0 thread started.
uart0_init() [OK]
RIOT network stack example application
All up, running the shell now
> ifconfig
ifconfig
Iface  7   HWaddr: da:5d:58:4d:66:16 
           inet6 addr: ff02::1/128  scope: local [multicast]
           inet6 addr: fe80::d85d:58ff:fe4d:6616/64  scope: local
           inet6 addr: ff02::1:ff4d:6616/128  scope: local [multicast]
           
> *** stack smashing detected ***: /home/kaspar/src/riot.8/examples/ng_networking/bin/native/ng_networking.elf terminated
Segmentation fault (core dumped)
[kaspar@booze ng_networking (benoit-timer)]$
```

master works fine (unless two packets arrive at the same time).

**Comment by benoit-canet** [comment]
The trampoline removal is the culprit.

I'll try without it.

**Comment by benoit-canet** [comment]
@kaspar030  Everything seems to work fine without the trampoline related commits. I will run the hwtimer_wait test a few hours and if it work I'll postpone my no assembly crusade for later ;)

**Comment by benoit-canet** [comment]
@kaspar030 works fine now including the ping6 -f test.

**Comment by kaspar030** [comment]
Works fine on OSX! Fast-ACK, I've cpmpiled native on linux and OSX (and there native didn't compile without).

**Comment by PeterKietzmann** [comment]
Ouh, what happened? Well, I'm really not inside this stuff. Just wanted to inform that the _tests/hwtimer_wait_ claims to take 13sec but actually takes ~20sec.

**Comment by benoit-canet** [comment]
takes 13s here.
Are you on OSX ?

**Comment by benoit-canet** [comment]
The test script use 20s

**Comment by benoit-canet** [comment]
I think your issue could be related to: https://github.com/RIOT-OS/RIOT/blob/master/cpu/native/periph/timer.c#L49

**Comment by benoit-canet** [comment]
It's needed to deal with clock skew but could delay the overall completion of the test

**Comment by benoit-canet** [comment]
@PeterKietzmann 

diff --git a/cpu/native/periph/timer.c b/cpu/native/periph/timer.c
index 3257e5e..ad5cf73 100644
--- a/cpu/native/periph/timer.c
+++ b/cpu/native/periph/timer.c
@@ -46,7 +46,7 @@
 
 #define NATIVE_TIMER_SPEED 1000000
 
-#define NATIVE_TIMER_MIN_RES 200
+#define NATIVE_TIMER_MIN_RES 1
 
 static unsigned long time_null;

The min res ajustement is the culprit and my subjective duration for 1 s is wrong.

We could apply this patch so we don't have any extra delay in the test but It means that something the hwtimer_wait test will fail because of clock skew.

**Comment by benoit-canet** [comment]
@PeterKietzmann I am testing a more sensible value: #define NATIVE_TIMER_MIN_RES 10 which does not affect the test duration time.

**Comment by benoit-canet** [comment]
@kaspar030 should we privilegiate precision over reliability ?

**Comment by thomaseichinger** [comment]
With this changes, `tests/vtimer_msg` stops after 2 to 6 seconds on OS X which wasn't the case before.

**Comment by PeterKietzmann** [comment]
@bapclenet good question.I think we should privilege reliability in general. Regarding the test, we can maybe adjust the application to a value that is reachable within a responsible resolution

**Comment by benoit-canet** [comment]
@thomaseichinger which change ? the whole pr or the small resolution patch in this discussion ?

**Comment by thomaseichinger** [comment]
@benoit-canet the whole PR.

**Comment by PeterKietzmann** [comment]
btw: with the smaller resolution the test takes ~15 sec.

**Comment by kaspar030** [comment]
@thomaseichinger What OS X version are you on? It doesn't stop here:

```Darwin milhouse.imp.fu-berlin.de 12.6.0 Darwin Kernel Version 12.6.0: Wed Mar 18 16:23:48 PDT 2015; root:xnu-2050.48.19~1/RELEASE_X86_64 x86_64```

**Comment by thomaseichinger** [comment]
Most recent OS X 10.10.4

```shell
$ uname -a
Darwin et-mbpr.local 14.4.0 Darwin Kernel Version 14.4.0: Thu May 28 11:35:04 PDT 2015; root:xnu-2782.30.5~1/RELEASE_X86_64 x86_64
```
@benoit-canet @kaspar030  I have to rephrase, it doesn't stop always but 19 from 20 times it stops either after 2 or 6 seconds.

**Comment by kaspar030** [comment]
@thomaseichinger Damn, it now hangs here, too. Wierd, I tried multiple times before merging.

**Comment by benoit-canet** [comment]
I am confused and annoyed by the fact that I do not have an OSx machine for debugging this.

**Comment by thomaseichinger** [comment]
There is also a missing include, taken care of in #3290 
```diff
diff --git a/cpu/native/periph/timer.c b/cpu/native/periph/timer.c
index 3257e5e..bdad39d 100644
--- a/cpu/native/periph/timer.c
+++ b/cpu/native/periph/timer.c
@@ -26,6 +26,7 @@
 #include <mach/clock.h>
 #include <mach/mach_init.h>
 #include <mach/mach_port.h>
+#include <mach/mach_host.h>
 #endif
 #include <time.h>
```

**Comment by kaspar030** [comment]
@thomaseichinger #3343

**Comment by kaspar030** [comment]
It doesn't always freeze...

**Comment by kaspar030** [comment]
Enabling debug in timer.c and hwtimer.c makes it freeze a lot less. ;(

**Comment by kaspar030** [comment]
The problem is vtimer setting an absolute timeout in the past.

**Comment by benoit-canet** [comment]
@thomaseichinger coud you try the following patch ?

diff --git a/cpu/native/include/native_internal.h b/cpu/native/include/native_internal.h                                                                                                                                                                                       
index e51ddea..382d81f 100644                                                                                                                                                                                                                                                  
--- a/cpu/native/include/native_internal.h                                                                                                                                                                                                                                     
+++ b/cpu/native/include/native_internal.h                                                                                                                                                                                                                                     
@@ -73,6 +73,8 @@ void _native_syscall_leave(void);                                                                                                                                                                                                                            
 void _native_syscall_enter(void);                                                                                                                                                                                                                                             
 void _native_init_syscalls(void);                                                                                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                                            
+void timer_clear_pending(void);                                                                                                                                                                                                                                                                                                                                                            
+                                                                                                                                                                                                                                                                                                                                                                                           
 /**                                                                                                                                                                                                                                                                                                                                                                                        
  * external functions regularly wrapped in native for direct use                                                                                                                                                                                                                                                                                                                           
  */                                                                                                                                                                                                                                                                                                                                                                                        
diff --git a/cpu/native/irq_cpu.c b/cpu/native/irq_cpu.c                                                                                                                                                                                                                                                                                                                                    
index 1a8ea8c..422026e 100644                                                                                                                                                                                                                                                                                                                                                               
--- a/cpu/native/irq_cpu.c                                                                                                                                                                                                                                                                                                                                                                  
+++ b/cpu/native/irq_cpu.c                                                                                                                                                                                                                                                                                                                                                                  
@@ -297,6 +297,10 @@ void native_isr_entry(int sig, siginfo_t *info, void *context)                                                                                                                                                                                                                                                                                                         
     _native_sigpend++;                                                                                                                                                                                                                                                                                                                                                                     
     //real_write(STDOUT_FILENO, "sigpend\n", 8);                                                                                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                                                                                                            
+    if (sig == SIGALRM) {                                                                                                                                                                                                                                                                                                                                                                  
+        timer_clear_pending();                                                                                                                                                                                                                                                                                                                                                             
+    }                                                                                                                                                                                                                                                                                                                                                                                      
+                                                                                                                                                                                                                                                                                                                                                                                           
     if (context == NULL) {                                                                                                                                                                                                                                                                                                                                                                 
         errx(EXIT_FAILURE, "native_isr_entry: context is null - unhandled");                                                                                                                                                                                                                                                                                                               
     }                                                                                                                                                                                                                                                                                                                                                                                      
diff --git a/cpu/native/periph/timer.c b/cpu/native/periph/timer.c                                                                                                                                                                                                                                                                                                                          
index 3257e5e..84e4bd9 100644                                                                                                                                                                                                                                                                                                                                                               
--- a/cpu/native/periph/timer.c                                                                                                                                                                                                                                                                                                                                                             
+++ b/cpu/native/periph/timer.c
@@ -54,6 +54,8 @@ static void (*_callback)(int);
 
 static struct itimerval itv;
 
+static volatile int pending;
+
 /**
  * returns ticks for give timespec
  */
@@ -83,6 +85,8 @@ int timer_init(tim_t dev, unsigned int ticks_per_us, void (*callback)(int))
         return -1;
     }
 
+    pending = 0;
+
     /* initialize time delta */
     time_null = 0;
     time_null = timer_read(0);
@@ -100,6 +104,10 @@ int timer_set(tim_t dev, int channel, unsigned int offset)
     (void)channel;
     DEBUG("%s\n", __func__);
 
+    if (offset && pending) {
+        timer_clear(dev, channel);
+    }
+
     if (offset && offset < NATIVE_TIMER_MIN_RES) {
         offset = NATIVE_TIMER_MIN_RES;
     }
@@ -116,9 +124,16 @@ int timer_set(tim_t dev, int channel, unsigned int offset)
     }
     _native_syscall_leave();
 
+    pending = !!offset;
+
     return 1;
 }
 
+void timer_clear_pending(void)
+{
+    pending = 0;
+}
+
 int timer_set_absolute(tim_t dev, int channel, unsigned int value)
 {
     uint32_t now = timer_read(dev);

**Comment by PeterKietzmann** [comment]
In that case the problem shouldn't be related to OSX right? _tests/vtimer_msg_ works for me. Is that still the misbehaving application you're talking about?

**Comment by kaspar030** [comment]
No, it's not vtimer. Seems like using "timer_set_absolute()" with very small offsets causes timer.c to underflow, e.g., set a hwtimer in the past.

**Comment by benoit-canet** [comment]
@kaspar030 that what NATIVE_TIMER_MIN_RES is suposed to prevent

**Comment by benoit-canet** [comment]
@kaspar030 but increasing it will slow down test/hwtimer_wait test further

**Comment by thomaseichinger** [comment]
@benoit-canet with `NATIVE_TIMER_MIN_RES` increased to 2000 it still freezes (sometimes)

**Comment by benoit-canet** [comment]
@thomaseichinger Can you try this patch ? http://paste.debian.net/280629/

**Comment by benoit-canet** [comment]
http://paste.debian.net/280646/

**Comment by benoit-canet** [comment]
@thomaseichinger Could you test this ? http://paste.debian.net/280649/

**Comment by benoit-canet** [comment]
@thomaseichinger and this http://paste.debian.net/280657/ to test @kaspar030 hypothesis

**Comment by gebart** [comment]
@kaspar030 Does #3147 help with the vtimer issue?

**Comment by thomaseichinger** [comment]
| patches | result |
| :--- | :--- |
| [1] | some improvement, i.e. less freezes |
| [1]+[2] | no further improvement |
| [1]+[2]+[3] | didn't see a freeze in 20 runs |

[1] http://paste.debian.net/280649/
[2] http://paste.debian.net/280657/
[3] #3147

**Comment by kaspar030** [comment]
Same here (didn't try 3).

**Comment by kaspar030** [comment]
[2] doesn't fix anything, delta calculation needs to be cast to (int32_t) to actually get a negative value.
(```int64_t delta = (int32_t)(value - now);)

**Comment by benoit-canet** [comment]
[2] should be delta <= 0 also to avoid cancelling the timer.

**Comment by benoit-canet** [comment]
@kaspar030 is it expected from the upper layer that setting a relative timer to 0 does a cancel ? Or should we rework timer.c ?

**Comment by kaspar030** [comment]
@benoit-canet No, setting to absolute 0 shouldn't cancel.

**Comment by benoit-canet** [comment]
Ok I'l prepare a series

**Comment by kaspar030** [comment]
I have a weird effect. If catching underflows like this:
```
int timer_set_absolute(tim_t dev, int channel, unsigned int value)
{
    uint32_t now = timer_read(dev);
    int64_t target = (int32_t)(value - now);

    DEBUG("timer_set_absolute(): delta=%lli\n", target);
    if (target < 0 && target > -NATIVE_TIMER_MIN_RES) {
        DEBUG("timer_set_absolute(): preventing underflow.\n");
        target = now + NATIVE_TIMER_MIN_RES;
    }   

    timer_set(dev, channel, target);

    return 1;
}
```

When the underflow happens, the test seems to delay exactly the amount of time timer_read() would return.

**Comment by OlegHahm** [comment]
> No, it's not vtimer. Seems like using "timer_set_absolute()" with very small offsets causes timer.c to underflow, e.g., set a hwtimer in the past.

Can't we just replace this call by `timer_set()` and get rid of this function completely? IIRC vtimer is the only user of this function and even vtimer is not really needing it.

**Comment by kaspar030** [comment]
wtimer also uses it, it's needed there to e.g., set a timer exactly to overflow. But wtimer checks a lot better for underflows.

**Comment by kaspar030** [comment]
@benoit-canet I have a patch ready to prevent the underflow.

**Comment by kaspar030** [comment]
See #3344. Together with [1], this makes the test run fine for me.

**Comment by benoit-canet** [comment]
@kaspar030 I did a pr for the zero value case

**Comment by kaspar030** [comment]
alright, @thomaseichinger, could you confirm that master + #3344 works for you?

**Comment by benoit-canet** [comment]
@kaspar030 any idea for the slowliness of the hwtimer_wait test case on native ? Should we define the idea of a minimal timer resolution in a central place and bake it in the hwtimer_wait/main.c loop test ?

**Comment by thomaseichinger** [comment]
@kaspar030 didn't see any freezes with #3344.

**Comment by PeterKietzmann** [comment]
@benoit-canet I stopped following your debugging steps here and in the other PRs so I can't comment in detail. In general I'd like to have an application that does what it says :-). If this is dependent on general timer specifications (like granularity), a global macro and a comment about the behaviour seem to be reasonable.

**Pull request merged**